woocommerce/woocommerce-android
# ExPlat integration

### ExPlat integration

Closes: 164-gh-woocommerce/woomobile-private


### Description

Integration with ExPlat will allow us to create experiments (e.g. A/B tests) in the project.

### Testing instructions

It's not possible to test the integration by itself since we don't have any experiments yet. 
To test if `ExPlat` instance is being provided properly:

1 - In the `ExperimentationModule`, inside the `provideExPlat` method, inject a mock `Experiment` in `experiments = setOf()`. e.g.:
```
experiments = setOf(object: Experiment {
            override val identifier: String = "test_experiment"
        })
```
2 - Inject `ExPlat` anywhere to test if the instance provided by `ExperimentationModule` is working. I've injected in `MainActivity` for testing. e.g.:
```
@Inject lateinit var exPlat: ExPlat
```

3 - Call `getVariation` using the same test `Experiment` identifier you've used in step `1`. e.g.
```
       val variation = exPlat.getVariation(
            object : Experiment {
                override val identifier = "test_experiment"
            }
        )
```
`variation` property `name` should be `control`:
<img width="853" alt="image" src="https://user-images.githubusercontent.com/14964993/165760384-e587dc63-b3fd-4286-a2e1-8ef30182818c.png">


4 - Call `getVariation` using a different identifier. e.g.
```
       val variation = exPlat.getVariation(
            object : Experiment {
                override val identifier = "test_experiment_2"
            }
        )
```

An exception should be thrown:
```
Caused by: java.lang.IllegalArgumentException: ExPlat: experiment not found: "test_experiment_2"! Make sure to include it in the set provided via constructor.
```

When we have real experiments, [here's an example](https://github.com/wordpress-mobile/WordPress-Android/pull/14286) by @renanferrari on how to test them using Charles proxy.

### Images/gif
N/A





Depends on https://github.com/Automattic/Automattic-Tracks-Android/pull/121
